lib/kube/proxy/forwarder.go

```diff
@@ -897,25 +904,23 @@ func wsProxy(wsSource *websocket.Conn, wsTarget *websocket.Conn) error {
 	return trace.Wrap(err)
 }
 
-// acquireConnectionLock acquires a semaphore used to limit connections to the Kubernetes agent.
+// AcquireConnectionLock acquires a semaphore used to limit connections to the Kubernetes agent.
 // The semaphore is releasted when the request is returned/connection is closed.
 // Returns an error if a semaphore could not be acquired.
-func (f *Forwarder) acquireConnectionLock(ctx context.Context, identity *authContext) error {
-	user := identity.Identity.GetIdentity().Username
-	roles, err := getRolesByName(f, identity.Identity.GetIdentity().Groups)
-	if err != nil {
-		return trace.Wrap(err)
+func (f *Forwarder) AcquireConnectionLock(ctx context.Context, user string, roles services.RoleSet) error {
+	maxConnections := roles.MaxKubernetesConnections()
+	if maxConnections == 0 {
+		return nil
 	}
 
-	maxConnections := services.RoleSet(roles).MaxKubernetesConnections()
 	semLock, err := services.AcquireSemaphoreLock(ctx, services.SemaphoreLockConfig{
```

Comment:
Currently, `services.AcquireSemaphoreLock`'s comment string says "`If successfully acquired,
background keepalive processes are started and an associated lock handle is returned.`", however it appears that that isn't actually happening in that function, and it's the caller's responsibility to call
```
go semLock.KeepAlive(ctx)
```

We should either change the comment or the behavior.

Reply:
Yeah, good point. I've amended `Acquire` to start the keepalive and made the keepalive function private.